### PR TITLE
[v8.0] feat: add documentation to setup pilots with tokens

### PR DESCRIPTION
**Pull request by aldbr**
Related to https://github.com/DIRACGrid/DIRAC/discussions/7123

BEGINRELEASENOTES
*docs
NEW: add documentation about setting up DIRAC to submit pilots with tokens
ENDRELEASENOTES

### CODE REVIEW

File: docs/source/AdministratorGuide/HowTo/pilotsWithTokens.rst [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+Setting up an `IdProvider`
+--------------------------
+
+- Set up an OAuth2 client in the token provider and obtain a `client_id` and a `client_secret`.
+
+    .. warning:: The client credentials obtained are confidential, store them in a secure place.
+       Any malicious user able to get access to them would be able to generate access tokens on your behalf.
+       To avoid any major issue, we recommend you to only grant essential privileges to the client (`compute` scopes).
+
+- Add the client credentials in your local `dirac.cfg` configuration such as::
```
Comment (andresailer) [inline docs/source/AdministratorGuide/HowTo/pilotsWithTokens.rst]:
```suggestion
- Add the client credentials in your local `dirac.cfg` configuration such as:
```
(Maybe also too much indent for the next block?)

Please have a look at https://dirac--7176.org.readthedocs.build/en/7176/AdministratorGuide/HowTo/pilotsWithTokens.html

File: docs/source/AdministratorGuide/HowTo/pilotsWithTokens.rst [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+- Connect the OIDC provider to a specific VO by adding `<IdProvider name>` in `/Registry/VO/<VO name>`.
+
+.. note:: Get more details about the DIRAC configuration from the :ref:`Configuration <dirac-configuration>` section.
+
+Launching the `TokenManagerHandler`
+-----------------------------------
+
+Run the following commands from a DIRAC client to install the `Framework/TokenManager` service:
+
+.. code-block:: bash
```
Comment (andresailer) [inline docs/source/AdministratorGuide/HowTo/pilotsWithTokens.rst]:
```suggestion
.. code-block:: console
```

File: docs/source/AdministratorGuide/HowTo/pilotsWithTokens.rst [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        IdProviders
+        {
+          <IdProvider name>
+          {
+            issuer = <OIDC provider issuer URL>
+          }
+        }
+      }
+
+- Connect the OIDC provider to a specific VO by adding `<IdProvider name>` in `/Registry/VO/<VO name>`.
```
Comment (andresailer) [inline docs/source/AdministratorGuide/HowTo/pilotsWithTokens.rst]:
```suggestion
- Connect the OIDC provider to a specific VO by adding the respective `<IdProvider name>` section under `/Registry/VO/<VO name>`.
```
If that is what is meant? Does it go under a `IdProviders` section as well?

File: docs/source/AdministratorGuide/HowTo/pilotsWithTokens.rst [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+Setting up an `IdProvider`
+--------------------------
+
+- Set up an OAuth2 client in the token provider and obtain a `client_id` and a `client_secret`.
+
+    .. warning:: The client credentials obtained are confidential, store them in a secure place.
+       Any malicious user able to get access to them would be able to generate access tokens on your behalf.
+       To avoid any major issue, we recommend you to only grant essential privileges to the client (`compute` scopes).
+
+- Add the client credentials in your local `dirac.cfg` configuration such as:
```
Comment (andresailer) [inline docs/source/AdministratorGuide/HowTo/pilotsWithTokens.rst]:
```suggestion
- Add the client credentials in the ``dirac.cfg`` of the relevant server configuration such as:
```
Right?

**Review by andresailer — COMMENTED** [review]
In principle almost all back ticks should probably be double backticks.
(Single backticks are references, double backticks are "code")

But is fine to me, except maybe for some clarifications.

Also maybe @marianne013 or @sfayer want to read through this as they have tried it?

File: docs/source/AdministratorGuide/HowTo/pilotsWithTokens.rst (hunk 3 above)
Comment (aldbr) [reply docs/source/AdministratorGuide/HowTo/pilotsWithTokens.rst]: No, I expanded that part with an example to make it more clear.

Comment (andresailer) [reply docs/source/AdministratorGuide/HowTo/pilotsWithTokens.rst]: Thanks, that is clear now.

**Comment by marianne013** [comment]
This needs a warning that this is not multi-VO compatible, i.e. for a given CE either all or none of the VOs using the CE have to use tokens.

**Comment by aldbr** [comment]
> given CE either all or none of the VOs using the CE have to use tokens.

I added a warning.
I am planning to take a look at the multi-VO case this week :crossed_fingers:

**Comment by marianne013** [comment]
Did we ever agree on any strategy for multi-VO ? I guess we could have a property of the CE along the line of X509VOs = ...., then as VOs transition to tokens there would be less and less of them in that list until the've all transitioned and it's empty.

**Comment by aldbr** [comment]
> Did we ever agree on any strategy for multi-VO ? 

Not really actually :sweat_smile: 

> I guess we could have a property of the CE along the line of X509VOs = ...., then as VOs transition to tokens there would be less and less of them in that list until the've all transitioned and it's empty.

But then you would have to manually update this list right?

What about having the logic in the `SiteDirector`. It would submit pilots with tokens under two conditions:
- (the current one) the CE is able to receive tokens.  Validation: `Tag = Token` should be included in the CE parameters.
- (a new one) the VO is able to produce tokens. Validation: `IdProvider` option is set in `/Registry/VO/<VO name>/`.

If one of these 2 conditions is not satisfied, then the Site Director submits pilots with proxies.

What do you think @marianne013 ?

**Comment by marianne013** [comment]
What if we have a VO and *some* sites use tokens and other don't ? I believe that is still the case for CMS (not that they are using DIRAC, but if CMS can't convince all their sites to update, what chance do I have ?) The transition will be a mess one way or the other. There is certainly something to be said for not trying tokens for a VO that doesn't have an non-X509 id provider.

**Comment by fstagni** [comment]
The `token` tag is per-CE. It basically says "this CE accepts tokens". I think your case might be that a certain CE might accept a token from VO "A" and proxy from VO "B".

**Comment by marianne013** [comment]
> The `token` tag is per-CE. It basically says "this CE accepts tokens". I think your case might be that a certain CE might accept a token from VO "A" and proxy from VO "B".

We assume that will be the most common scenario, not all VOs will get their tokens ready at the same time.

**Comment by aldbr** [comment]
With the solution I proposed:
- VO "A" would be able to submit pilots with tokens
- VO "B", which is not yet ready, would not set the `IdProvider` parameter and would still be able to submit with proxies to the same CE (SiteDirector would see that the CE is ready for tokens, but not the VO).

The only problem I see with the solution I proposed is the transition period. Here is an example:
- VO "A" has convinced Site1, Site2 to accept their tokens. CEs from Site1, Site2 would have `Tag = Token`.
- VO "B" is transitioning to tokens. Site1 accepts their tokens but Site2 needs some time to adjust their configuration (for some obscure reasons). In such a case, the Site Director would try to submit pilots with tokens to Site2, which would not be accepted.

Do you think it would be a problem?

**Comment by sfayer** [comment]
> The only problem I see with the solution I proposed is the transition period....
> Do you think it would be a problem?

Hi,

Unfortunately I think this is likely to be a problem: It's likely that we will want to switch VOs over to tokens one at a time, test it on a small number of CE/Sites for a while and then scale it up to other sites (e.g. a long transition period). Would it be possible to add a setting or something to allow us to accommodate that?

Regards,
Simon

File: docs/source/AdministratorGuide/HowTo/pilotsWithTokens.rst [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+Launching the ``TokenManagerHandler``
+-------------------------------------
+
+Run the following commands from a DIRAC client to install the ``Framework/TokenManager`` service:
+
+.. code-block:: console
+
+   $ dirac-proxy-init -g dirac_admin
+
+   $ dirac-admin-sysadmin-cli --host <dirac host>
```
Comment (aldbr) [inline docs/source/AdministratorGuide/HowTo/pilotsWithTokens.rst:76]: Add a comment: `The TokenManager runs with Tornado`

**Comment by aldbr** [comment]
You can actually play with the `VO` option defined in the Sites/CEs/Queues.
Example:

```
CE1
{
  VO = vo1, vo2, vo3, vo4
  ...
}
```

Could become:

```
CE1_proxy
{
  VO = vo1, vo2
}
CE1
{
  VO = vo3, vo4
  Tag = Token
}
```

This would be probably more consistent to work with the existing options than having a new option where you can actually define anything (e.g. `VO = vo1, vo2` whereas you have `X509VOs = vo1, vo3`).

**Comment by aldbr** [comment]
This PR can be merged once https://github.com/DIRACGrid/DIRAC/pull/7208 is merged.
I added some documentation for multi-VO.

**Review by fstagni — APPROVED** [review]

**Pull request merged**